### ISSUE TITLE
null stimulus duration in QC passed sweeps

### ISSUE THREAD

**Issue by tmchartrand** [AllenInstitute/ipfx#525]
Moving the discussion from email here so we have a record moving forward.

After the recent fixes, cells are failing Ephys QC v3 due to null stimulus_duration (not allowed by the QcSweepFeatures schema).
Null stimulus amplitude is actually an intended output from get_stim_characteristics for a few various error cases, but these sweeps should already be tagged as incomplete and thus not passed to the QC stage.

From @ru57y34nn: "I noticed that some sweep have null stimulus_duration but do not cause issues if the sweep is tagged (e.g. “Recording stopped before completing the experiment epoch”) but it will cause an issue if the stimulus_duration is null and the sweep is not tagged. "

The plot below shows an nwb with both type of sweeps: the one that stops recording right after stimulus onset is tagged, while the one that stops mid-stimulus is not. Both sweeps have null duration. The correct fix here is for the epoch code to give both cases null stimulus epochs, so the sweeps are tagged/dropped correctly. I'll put a PR up for this shortly (I think we actually had added an equivalent fix in @ru57y34nn's original epoch fix PR and then I forgot to add it to my version).

![image](https://user-images.githubusercontent.com/12821536/129283143-054d49de-14cf-4cde-956f-6c2379bd8351.png)

This does bring back up the inconsistency of allowing null for stimulus_duration (in LIMS, but not in QC input) while not allowing null for stimulus_amplitude (discussed in #498 / #503). Perhaps it's worth revisiting that and allowing null amplitude (@MatthewAitken)?

**Comment by MatthewAitken** [comment]
> This does bring back up the inconsistency of allowing null for stimulus_duration (in LIMS, but not in QC input) while not allowing null for stimulus_amplitude (discussed in #498 / #503). Perhaps it's worth revisiting that and allowing null amplitude (@MatthewAitken)?

I agree - I prefer NaNs/nulls when it truly is NaN and not zero. I'll ask the Platform team about the possibility of updating this.

**Comment by timjarsky** [comment]
@tmchartrand the plot in the lower right (ramp stimulus) looks good to me. -  I'm expecting many ramps to look like that. 

Are you saying they won't be passed on to the QC stage and ultimately feature extraction?

**Comment by tmchartrand** [comment]
@timjarsky no I should have specified sweep numbers - I'm talking sweeps 20 and 58, both long squares.

**Comment by timjarsky** [comment]
> @timjarsky no I should have specified sweep numbers - I'm talking sweeps 20 and 58, both long squares.

gotcha. thanks for the clarifcation.

**Comment by timjarsky** [comment]
@tmchartrand we recently merged a PR in MIES that documents all known reasons for early termination https://github.com/AllenInstitute/MIES/pull/943

**Comment by ru57y34nn** [comment]
Thanks @tmchartrand for chasing down the issue and providing the fix!
